src/core/server/status/plugins_status.ts

```diff
@@ -57,114 +97,139 @@ export class PluginsStatusService {
+          pluginData[dependency].reverseDependencies.push(name);
+        });
+      });
     }
 
-    return this.update$.pipe(
-      switchMap(() => {
-        const pluginStatuses = plugins
-          .map((depName) => {
-            const pluginStatus = this.pluginStatuses.get(depName)
-              ? this.pluginStatuses.get(depName)!.pipe(
-                  timeoutWith(
-                    STATUS_TIMEOUT_MS,
-                    this.pluginStatuses.get(depName)!.pipe(
-                      startWith({
-                        level: ServiceStatusLevels.unavailable,
-                        summary: `Status check timed out after ${STATUS_TIMEOUT_MS / 1000}s`,
-                      })
-                    )
-                  )
-                )
-              : this.getDerivedStatus$(depName);
-            return [depName, pluginStatus] as [PluginName, Observable<ServiceStatus>];
-          })
-          .map(([pName, status$]) =>
-            status$.pipe(map((status) => [pName, status] as [PluginName, ServiceStatus]))
-          );
-
-        return combineLatest(pluginStatuses).pipe(
-          map((statuses) => Object.fromEntries(statuses)),
-          distinctUntilChanged<Record<PluginName, ServiceStatus>>(isDeepStrictEqual)
-        );
-      })
+    return pluginData;
+  }
+
+  private getRootPlugins(): PluginName[] {
+    return Object.keys(this.pluginData).filter(
+      (plugin) => this.pluginData[plugin].dependencies.length === 0
+    );
+  }
+
+  private getOrderedPluginNames(): PluginName[] {
+    this.rootPlugins.forEach((plugin) => {
+      this.calculateDepthRecursive(plugin, 1);
+    });
+
+    return sortBy(Object.values(this.pluginData), ['depth', 'name']).map(({ name }) => name);
+  }
+
+  private calculateDepthRecursive(plugin: PluginName, depth: number) {
+    const pluginData = this.pluginData[plugin];
+    pluginData.depth = Math.max(pluginData.depth, depth);
+    const newDepth = depth + 1;
+    pluginData.reverseDependencies.forEach((revDep) =>
+      this.calculateDepthRecursive(revDep, newDepth)
     );
   }
 
-  /**
-   * Determines whether or not this plugin or any plugin in it's dependency tree have a custom status registered.
-   */
-  private anyCustomStatuses(plugin: PluginName): boolean {
-    if (this.pluginStatuses.get(plugin)) {
-      return true;
+  private updatePluginStatus(plugin: PluginName): void {
+    const newStatus = this.determinePluginStatus(plugin);
+    const pluginData = this.pluginData[plugin];
+    pluginData.derivedStatus = newStatus;
+
+    if (!pluginData.reportedStatus) {
+      // this plugin has NOT reported any status yet. Thus, its status is derived from its dependencies + core
+      this.pluginStatus[plugin] = newStatus;
+    }
+  }
+
+  private updatePluginsStatuses(plugins: PluginName[]): void {
+    const toCheck = new Set<PluginName>(plugins);
+    for (let i = 0; i < this.orderedPluginNames.length && toCheck.size > 0; ++i) {
```

Comment:
question: maybe we should iterate through `toCheck` set?

Reply:
Nope, because we need to iterate in "depth" order, aka first updating root plugins' statuses, then updating `depth=2`, and so on so forth. Otherwise we might try to update a plugin status before updating its dependencies.